packages/flutter/lib/src/widgets/restoration_properties.dart

```diff
@@ -143,44 +143,50 @@ abstract class RestorableValue<T> extends RestorableProperty<T> {
   void didUpdateValue(T? oldValue);
 }
 
-// _RestorablePrimitiveValue and its subclasses do not allow null values in
-// anticipation of NNBD (non-nullability by default).
-//
-// If necessary, we can in the future define a new subclass hierarchy that
-// does allow null values for primitive types. Borrowing from lisp where
-// functions that returned a bool ended in 'p', a suggested naming scheme for
-// these new subclasses could be to add 'N' (for nullable) to the end of a
-// class name (e.g. RestorableIntN, RestorableStringN, etc.) to distinguish them
-// from their non-nullable friends.
-class _RestorablePrimitiveValue<T extends Object> extends RestorableValue<T> {
-  _RestorablePrimitiveValue(this._defaultValue)
-    : assert(_defaultValue != null),
-      assert(debugIsSerializableForRestoration(_defaultValue)),
+// _RestorablePrimitiveValueN and its subclasses allows for null values.
+// See [_RestorablePrimitiveValue] for the non-nullable version of this class.
+class _RestorablePrimitiveValueN<T extends Object?> extends RestorableValue<T> {
+  _RestorablePrimitiveValueN(this._defaultValue)
+    : assert(debugIsSerializableForRestoration(_defaultValue)),
       super();
 
   final T _defaultValue;
 
   @override
   T createDefaultValue() => _defaultValue;
 
-  @override
-  set value(T value) {
-    assert(value != null);
-    super.value = value;
-  }
-
   @override
   void didUpdateValue(T? oldValue) {
     assert(debugIsSerializableForRestoration(value));
     notifyListeners();
   }
 
+  @override
+  T fromPrimitives(Object? serialized) => serialized as T;
+
+  @override
+  Object? toPrimitives() => value;
+}
+
+// _RestorablePrimitiveValue and its subclasses do not allow null values in
+// anticipation of NNBD (non-nullability by default).
+class _RestorablePrimitiveValue<T extends Object> extends _RestorablePrimitiveValueN<T> {
+  _RestorablePrimitiveValue(T _defaultValue)
+    : assert(_defaultValue != null),
```

Comment:
Are the asserts here necessary anymore, since non-nullability is defined as part of the class? It shouldn't be possible to pass a null value in to begin with.

Reply:
Hm, I guess I can't think of a way to write a test for this to prevent a regression if someone were to make it nullable by accident, so the asserts act as sort of a line of defense?

Reply:
Eventually, we can remove these asserts. For now we are keeping them for people who are not running their app with null-safety enabled.